zebra-chain/src/note_encryption.rs

```diff
@@ -0,0 +1,90 @@
+//! Note encryption types.
+
+use std::{
+    convert::{AsRef, From},
+    fmt,
+};
+
+mod sapling;
+mod sprout;
+
+/// A 512-byte _Memo_ field associated with a note, as described in
+/// [protocol specification §5.5][ps].
+///
+/// The usage of the memo field is by agreement between the sender and
+/// recipient of the note.
+///
+/// [ps]: https://zips.z.cash/protocol/protocol.pdf#spendencoding
+#[derive(Clone, Copy)]
+pub struct Memo([u8; 512]);
+
+impl<T: AsRef<[u8]>> From<T> for Memo {
+    fn from(input: T) -> Self {
+        let input_bytes: &[u8] = input.as_ref();
+
+        let mut full_bytes = [0; 512];
+
+        if input_bytes.len() < 512 {
+            full_bytes[0..input_bytes.len()].copy_from_slice(input_bytes);
+            Memo(full_bytes)
+        } else if input_bytes.len() == 512 {
+            full_bytes[..].copy_from_slice(input_bytes);
+            Memo(full_bytes)
+        } else {
+            // Because this is a From impl, we truncate your input
+            // rather than return any Memo at all.
```

Comment:
This might be a good reason to use `TryFrom` instead of `From`? It seems like a potential sharp edge to silently truncate the memo contents.

Reply:
Unfortunately I run into this 

https://github.com/rust-lang/rust/issues/50133

Reply:
Ah. Yeah that's annoying. In that case I think that it's less bad to do
```
impl<'a> TryFrom<&'a [u8]> for Memo {}
```
than to keep the `AsRef` ergonomics but lose the fallibility.

Reply:
The reason being that a caller with `T: AsRef<[u8>` can always do `.as_ref()` to get a slice and then use the slice-specific impl, but can't work around the lack of fallibility.